### PR TITLE
[GEOS-11275] Wicket 9 upgrade

### PR DESCRIPTION
**Pull request by bradh**
[GEOS-11275](https://osgeo-org.atlassian.net/browse/GEOS-11275) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

Work in progress on Wicket 8 migration (https://github.com/geoserver/geoserver/wiki/Jakarta-EE#wicket)

Here is content to fix found during testing:

* [Wicket Test Plan](https://docs.google.com/spreadsheets/d/1pQmncG4zxpgJnHxeI4myFfOBD17U2CIMcy59II4XAfo/edit?usp=sharing) (Google Doc)

* Look for "TODO WICKET8" for remaining work. Mostly this is verification, which should occur at the end. However there are broken modules and broken tests that need to be debugged.

# Checklist

- [ ] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [ ] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by aaime** [comment]
@bradh I was thinking, if most of the remaining work is to interactively test stuff, maybe we could try to involve others as well (e.g., users) just just asking to run side by side a GeoServer 2.24.x and a binary from this branch? Maybe setup a spreadsheet of pages to test to divide work. Just thinking out loud.

**Comment by bradh** [comment]
There is still one part that I need to fix: https://github.com/geoserver/geoserver/pull/7154/files#diff-150ffebe59f506d34ef00088e4dbe5e57cb4372a806a40c249ce7f0b00c3aca1

I am currently thinking something like:

```
diff --git a/src/web/security/core/src/test/java/org/geoserver/security/web/AbstractSecurityNamedServicePanelTest.java b/src/web/security/core/src/test/java/org/geoserver/security/web/AbstractSecurityNamedServicePanelTest.java
index e135f31965..268e9fa4cb 100644
--- a/src/web/security/core/src/test/java/org/geoserver/security/web/AbstractSecurityNamedServicePanelTest.java
+++ b/src/web/security/core/src/test/java/org/geoserver/security/web/AbstractSecurityNamedServicePanelTest.java
@@ -14,9 +14,12 @@
 import java.util.List;
 import org.apache.wicket.Component;
 import org.apache.wicket.ajax.AjaxRequestHandler;
+import org.apache.wicket.ajax.markup.html.AjaxLink;
 import org.apache.wicket.extensions.ajax.markup.html.modal.ModalWindow;
 import org.apache.wicket.markup.html.form.CheckBox;
 import org.apache.wicket.markup.html.form.FormComponent;
+import org.apache.wicket.markup.html.list.ListItem;
+import org.apache.wicket.markup.html.list.ListView;
 import org.apache.wicket.markup.repeater.Item;
 import org.apache.wicket.markup.repeater.data.DataView;
 import org.apache.wicket.util.tester.FormTester;
@@ -203,27 +206,24 @@ protected String getSecurityConfigClassName() {
         return formTester.getForm().get("config.className").getDefaultModelObjectAsString();
     }
 
-    @SuppressWarnings("deprecation")
     protected <T extends SecurityNamedServicePanelInfo> void setSecurityConfigClassName(
-            Class<T> clazz) {
-        // TODO WICKET8 - This should not be commented out.
-        // ListView list = (ListView)
-        // tester.getLastRenderedPage().get("servicesContainer:services");
-        // int toClick = -1;
-        // for (int i = 0; i < list.getList().size(); i++) {
-        //     if (clazz.isInstance(list.getList().get(i))) {
-        //        toClick = i;
-        //        break;
-        //    }
-        // }
-        // AjaxLink link = (AjaxLink) list.get(toClick).get("link");
-        // if (link.isEnabled()) {
-        //    tester.executeAjaxEvent(link, "click");
-        // }
-        //        formTester.select("config.className", index);
-        //
-        // tester.executeAjaxEvent(formTester.getForm().getPageRelativePath()+":config.className",
-        // "change");
+            Class<T> clazz) throws Exception {
+        ListView list = (ListView) tester.getLastRenderedPage().get("servicesContainer:services");
+        list.forEach(
+                i -> {
+                    if (i instanceof ListItem) {
+                        ListItem listItem = (ListItem) i;
+                        listItem.forEach(
+                                action -> {
+                                    if (action instanceof AjaxLink) {
+                                        AjaxLink link = (AjaxLink) action;
+                                        if (link.isEnabled()) {
+                                            tester.executeAjaxEvent(link, "click");
+                                        }
+                                    }
+                                });
+                    }
+                });
     }
 
     protected void clickSave() {
```

but not sure I understand the base code that well, or how to adequately test this.

In terms of user testing, maybe that would be best saved for post- Wicket 9? Otherwise we'll just be going back to ask for more test help pretty soon.

For Wicket 9, I think the biggest work is going to be https://github.com/geoserver/geoserver/pull/7154/files#diff-0f799f2da8c62ce7254c4f4c914fd1f0ad5b3968a3bbc73c0fda75e7f458fe68
Any chance you want to tackle that?

**Comment by aaime** [comment]
Good idea to wait for wicket 9 as well before involving users (not sure you wanted to do that too given the PR title, but good we're moving there as well).

The link you referred me to shows 229 modified files, not sure which is the specific issue. But yeah, happy to help in general, just mind I might take a long time if it's big.

File: src/web/core/src/main/java/org/geoserver/web/wicket/CodeMirrorEditor.java [hunk 1]
```diff
@@ -45,6 +41,9 @@
  *
  * @author Andrea Aime
  */
+// TODO WICKET8 - migrate to https://yauaa.basjes.nl/ per
```
Comment (bradh) [inline src/web/core/src/main/java/org/geoserver/web/wicket/CodeMirrorEditor.java:44]: @aaime This is the codemirror part.

**Comment by bradh** [comment]
> Good idea to wait for wicket 9 as well before involving users (not sure you wanted to do that too given the PR title, but good we're moving there as well).

I think I'm well advanced on wicket 9 now. At least it builds and passes CI. A lot of deprecation suppressions for the ModalWindow change (which will block Wicket 10), and haven't tested the runtime impact of the CSP policy yet.

> The link you referred me to shows 229 modified files, not sure which is the specific issue. But yeah, happy to help in general, just mind I might take a long time if it's big.

I hoped it would go to a file within that list. The problem is the codemirror editor, which uses functions that are deprecated in wicket 8 and removed in wicket 9, to do sniffing of browser versions. Have added a comment that should take you to that.

**Comment by aaime** [comment]
Regarding the diff in AbstractSecurityNamedServicePanelTest, I had a look at the code. I don't remember working on it before, but semantically, the new bits are not doing the same thing. The old code was getting a ListView, was checking its contents, and if the item in it was matching a given reference class, it was extracting the ajax links, and clicking it.
The current code just iterates over all the the elements in the list, finds all ajax links in them, and clicks each one of them.
It might incidentally end up getting the same result (only one enabled ajax link anyways), but it does not seem to be the same in general.

**Comment by aaime** [comment]
Wow, this browser detection thing is a real can of work. I get that the old Wicket code might have been imprecise, but Yauaa is really overboard:

> To actually use it in your application you need to create an instance of the UserAgentAnalyzer.
> 
> Please instantiate a new UserAgentAnalyzer as few times as possible because the initialization step for a full UserAgentAnalyzer (i.e. all fields) usually takes something in the range of 2-5 seconds and uses a few hundred MiB of memory to store all the analysis datastructures.
>
> This analyzer can only do a single analysis at a time and to ensure correct working the main method is synchronized

We cannot seriously keep a pool of these objects around, each one with a monster "few hundred megabytes"... need to think of some alternative.

**Comment by aaime** [comment]
[ua-parser](https://github.com/ua-parser/uap-java) seems promising size wise, but appears to have been abandoned months ago... it depends on a version of snakeyaml that has vulnerabilties, and PRs to upgrade Snakeyaml or just remove the dependency have been ignored.

**Comment by aaime** [comment]
Perhaps a simpler approach is to just assume that in 2023, no one should be using IE 8, Firefox less than 3, Safari less than 5, or Opera less than 9. Which seems to be a fair assumption based on the browser version stats in these page:

https://www.stetic.com/market-share/browser/
https://caniuse.com/usage-table

**Comment by bradh** [comment]
> Perhaps a simpler approach is to just assume that in 2023, no one should be using IE 8, Firefox less than 3, Safari less than 5, or Opera less than 9. Which seems to be a fair assumption based on the browser version stats in these page:

Or if they are, they won't be upgrading geoserver. Its possible that those browsers will have other issues with wicket 10 features anyway.

So the replacement is just removal of the detection code.

**Comment by aaime** [comment]
The real "fun" seems to be replacing ModalWindow with ModalDialog ... as far as I can tell, it's more primitive but also more flexible. There is an example of how to rebuild a modal window like experience here:
https://github.com/apache/wicket/tree/master/wicket-examples/src/main/java/org/apache/wicket/examples/ajax/builtin/modal

That is going to affect quite a bit of places:

```
aaime@colossus ~/devel/git-gs (wicket8) $ git grep -l "new ModalWindow" | grep -v Test   | sort | uniq
src/community/cog/src/main/java/org/geoserver/web/data/store/cog/panel/CogInput.java
src/community/elasticsearch/src/main/java/org/geoserver/elasticsearch/ElasticConfigurationPanel.java
src/community/solr/src/main/java/org/geoserver/solr/SolrConfigurationPanel.java
src/extension/importer/web/src/main/java/org/geoserver/importer/web/ImportTaskTable.java
src/extension/metadata/src/main/java/org/geoserver/metadata/web/panel/ProgressPanel.java
src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/ComplexInputPanel.java
src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/WPSRequestBuilder.java
src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/WPSRequestBuilderPanel.java
src/web/core/src/main/java/org/geoserver/web/admin/ModuleStatusPanel.java
src/web/core/src/main/java/org/geoserver/web/data/layergroup/LayerGroupEntryPanel.java
src/web/core/src/main/java/org/geoserver/web/data/layergroup/RootLayerEntryPanel.java
src/web/core/src/main/java/org/geoserver/web/data/resource/FeatureResourceConfigurationPanel.java
src/web/core/src/main/java/org/geoserver/web/data/store/StorePanel.java
src/web/core/src/main/java/org/geoserver/web/wicket/browser/FileInput.java
src/web/core/src/main/java/org/geoserver/web/wicket/CRSPanel.java
src/web/core/src/main/java/org/geoserver/web/wicket/GeoServerDialog.java
src/web/demo/src/main/java/org/geoserver/web/demo/DemoRequestsPage.java
src/web/demo/src/main/java/org/geoserver/web/demo/ReprojectPage.java
src/web/wcs/src/main/java/org/geoserver/wcs/web/demo/WCSRequestBuilder.java
src/web/wcs/src/main/java/org/geoserver/wcs/web/demo/WCSRequestBuilderPanel.java
src/web/wms/src/main/java/org/geoserver/wms/web/data/AbstractStylePage.java
src/web/wms/src/main/java/org/geoserver/wms/web/WMSAdminPage.java
```

I'm thinking we could rebuild our own ModalWindow on top of ModalDialog, in a way that's mostly API compatible with the old deprecated class, at least for the few parts that we actually use.

**Comment by bradh** [comment]
> Regarding the diff in AbstractSecurityNamedServicePanelTest, I had a look at the code. I don't remember working on it before, but semantically, the new bits are not doing the same thing. The old code was getting a ListView, was checking its contents, and if the item in it was matching a given reference class, it was extracting the ajax links, and clicking it. The current code just iterates over all the the elements in the list, finds all ajax links in them, and clicks each one of them. It might incidentally end up getting the same result (only one enabled ajax link anyways), but it does not seem to be the same in general.

I've reverted this.

**Comment by bradh** [comment]
> I'm thinking we could rebuild our own ModalWindow on top of ModalDialog, in a way that's mostly API compatible with the old deprecated class, at least for the few parts that we actually use.

I've just pushed a commit that centralises the use of ModalWindow into a GSModalWindow wrapper.
The wrapper is here:
https://github.com/geoserver/geoserver/pull/7154/commits/d82bc2b22e228e60e88b7d8b3077c8fc83444f90#diff-d848262a28b51311cd5d074fc78d30bc4c6c16e32f87be5c75d26dcf70d3a090

Based on that, it might be worth doing a couple of implementations (possibly with a shared base class if we need to share styling). One case is the GeoServerDialog, which is the only place that uses a bunch of the GSModalWindow methods. Then the other implementation is the "everything else" case.  A later refinement could be to create a GSModalPopup, since there is a bit of duplicated (probably copy-n-pasted) code spread across multiple packages for that.

**Comment by aaime** [comment]
> I've just pushed a commit that centralises the use of ModalWindow into a GSModalWindow wrapper. 

One deprecation suppression to rule them all! :tada:

**Comment by bradh** [comment]
> > Perhaps a simpler approach is to just assume that in 2023, no one should be using IE 8, Firefox less than 3, Safari less than 5, or Opera less than 9. Which seems to be a fair assumption based on the browser version stats in these page:

> So the replacement is just removal of the detection code.

This is implemented in https://github.com/geoserver/geoserver/pull/7154/commits/801f68253544f6c929168d74b6bae331bd55044e

**Comment by bradh** [comment]
Wicket 9 has a strict (perhaps aggressive would be a better term) default policy for Content Security Policy (CSP) - see https://github.com/apache/wicket/blob/master/wicket-user-guide/src/main/asciidoc/security/security_6.adoc

That is currently completely disabled by https://github.com/geoserver/geoserver/pull/7154/commits/c16c7526f8a17dc2b8b202800bab86168d031b9d 

I'd like to "do it right", which will be a bit of work and the manual testing should come after that. Mitigating more XSS and data injection attacks is worth it.

**Comment by bradh** [comment]
Problems to fix:

 - [ ] Opening logs, produces: "Uncaught TypeError: textArea is null
    <anonymous> http://localhost:8080/geoserver/web/wicket/bookmarkable/org.geoserver.web.admin.LogPage?8&filter=false&filter=false:42"

- [ ] Checkboxes can not be toggled on Image Processing page

**Comment by aaime** [comment]
Is this good for a manual test round, involving users?

**Comment by bradh** [comment]
> Is this good for a manual test round, involving users?

I think its premature, but will need help to bash it into shape. Will follow-up on the mailing list with details.

File: src/web/core/src/main/java/org/geoserver/web/GeoServerApplication.java [hunk 2]
```diff
@@ -220,8 +222,26 @@ protected void init() {
...
+                // TODO: we definitely want to remove UNSAFE_INLINE and UNSAFE_EVAL
+                .add(
+                        CSPDirective.SCRIPT_SRC,
+                        CSPDirectiveSrcValue.SELF,
+                        CSPDirectiveSrcValue.UNSAFE_INLINE)
+                // TODO: we definitely want to  remove UNSAFE_INLINE
+                .add(
+                        CSPDirective.STYLE_SRC,
+                        CSPDirectiveSrcValue.SELF,
+                        CSPDirectiveSrcValue.UNSAFE_INLINE)
```
Comment (bradh) [inline src/web/core/src/main/java/org/geoserver/web/GeoServerApplication.java:238]: This is the line to remove to see problems with inline styling.

**Comment by aaime** [comment]
@bradh several inline styling issues are due to ModalWindow adding "display:none" in its "onComponentTag" method, which triggers CSP. I've fixed one case in GeoServerDialog, but I guess we'd need the GSModalWindow experiment you cited above to centralize this change in one place.

**Comment by jodygarnett** [comment]
@bradh I could not find a Jira issue for this so I am making one; enough people would like to help out we need some place to organize communication.

**Comment by bradh** [comment]
> @bradh I could not find a Jira issue for this so I am making one; enough people would like to help out we need some place to organize communication.

@jodygarnett Thanks

**Comment by petersmythe** [comment]
Please also see the wiki page: https://github.com/geoserver/geoserver/wiki/Upgrade-to-Apache-Wicket-10 as maybe a place to coordinate efforts.

**Comment by bradh** [comment]
I don't really need much more management oversight,  but rather PRs to get stuff ready for test.

**Comment by bradh** [comment]
Nasty rebase....

**Comment by jodygarnett** [comment]
@bradh I do not have time to write a blog post; but GeoCat is in position to handle the A/B testing to confirm the pages work (when this is ready for manual testing).

**Comment by jodygarnett** [comment]
@DBlasby I have attempted to merge main onto this branch for a PR....

But it understandably conflicts due the changes made to the demo request page. I expect it may be easier to reconvert this section to wicket 8 given the extensive changes.

**Comment by davidblasby** [comment]
@bradh 

I created a PR against your branch - 

https://github.com/bradh/geoserver/pull/16

Changes for content-security-policy for:

GS Main Page
Demo Requests
WCS Request Builder
WPS Request Builder

CF the PR and the message to geoserver-devel.  The changes are pretty simple, but please review.

**Comment by davidblasby** [comment]
I am going to look at few other pages - mostly the jwt-headers community security module (which likely will be having CSP issues).

@MichelGabriel reported some CSP warnings on these pages:

* Image Processing (element event handler)
* Disk Quota (inline style)
* Passwords (inline style)
* Data Security (inline style)

**Comment by jodygarnett** [comment]
Thanks @davidblasby and @MichelGabriel - if we could get this in for the 2.26 release cycle it would take some risk out the [spring-framework 6](https://geoserver.org/behind%20the%20scenes/2024/01/03/roadmap.html) roadmap.

**Comment by davidblasby** [comment]
@bradh - I created a PR to your branch
https://github.com/bradh/geoserver/pull/17

This contains CSP fixes for the Security Section.  I think this fixes all the "standard" (i.e. no extensions or community modules).  Just about to work on jwt-headers.

**Comment by davidblasby** [comment]
@bradh - another PR.  This is for JWT Headers.

https://github.com/bradh/geoserver/pull/18

**Comment by davidblasby** [comment]
Status (GS left column headers):

* [x] About & Status
* [ ] Data
* [ ] Services
* [ ] Settings
* [ ] Tile Caching
* [x] Security ("`release`")
* [x] Demos ("`release`")
* [x] Tools ("`release`")
* [ ] Community Modules
     * [x] JWT Headers
     * [ ] Oauth-OpenID-Connect

**Comment by MichelGabriel** [comment]
I've visually compared this PR with the main branch and found some small interface issues:

### Data Security page

The radio buttons have the labels on a new line
gs-data

### Style Editor

The editor has no border and the 2 blocks, `Style Data` and `Legend` are very close to each other.
gs-styleditor

### Demo requests

The editor has no border
gs-demorequest

**Comment by MichelGabriel** [comment]
I've visually reviewed this PR with the main branch. I've found 2 issues. I will make a small PR to fix them.

### Style Editor

The editor has no border and the 2 blocks, `Style Data` and `Legend` are very close to each other.
gs-styleditor

### Demo requests

The editor has no border
gs-demorequest

**Comment by jodygarnett** [comment]
@davidblasby wrote:

> Please fill in what needs to be checked.

@MichelGabriel wrote:

> I've visually reviewed this PR with the main branch.

Can you two sort out what is left, then we can take this PR out of draft.

**Comment by davidblasby** [comment]
@bradh - I've noticed some problems with the `GSModalWindow` (which wraps a wicket `ModalWindow`).

Inside the browser, wicket will create an element like this;

```
<div class="wicket-modal" id="_wicket_window_78" role="dialog" aria-labelledby="_wicket_window_90" style="top: 10px; left: 10px; width: 100px;">
   <form style='background-color:transparent;padding:0px;margin:0px;border-width:0px;position:static'>
      <div id="_wicket_window_79">
         <div class="w_top_1">
            <div class="w_topLeft" id="_wicket_window_83"></div>
            <div class="w_topRight" id="_wicket_window_84"></div>
            <div class="w_top" id='_wicket_window_82'></div>
         </div>
         <div class="w_left" id='_wicket_window_85'>
            <div class="w_right_1">
               <div class="w_right" id='_wicket_window_86'>
                  <div class="w_content_1">
                     <div class="w_caption"  id="_wicket_window_80">
                        <a class="w_close" style="z-index:1" href="#"></a>
                        <h3 id="_wicket_window_90" class="w_captionText"></h3>
                     </div>
                     <div class="w_content_2">
                        <div class="w_content_3">
                           <div class="w_content">
                              <div id='_wicket_window_81' class='w_content_container'></div>
                           </div>
                        </div>
                     </div>
                  </div>
               </div>
            </div>
         </div>
         <div class="w_bottom_1" id="_wicket_window_89_1">
            <div class="w_bottomRight"  id="_wicket_window_88"></div>
            <div class="w_bottomLeft" id="_wicket_window_87"></div>
            <div class="w_bottom" id="_wicket_window_89"></div>
         </div>
      </div>
   </form>
</div>
```

This is created here at line 127: http://localhost:8080/geoserver/web/wicket/resource/org.apache.wicket.extensions.ajax.markup.html.modal.ModalWindow/res/modal-ver-329D3142F0E943590C49E1230F02B47E.js 

This contains some `style="..."` tags.  This, of course, causes CSP warning.  

I'm not sure how to fix fix this w/o modifying the Wicket JS code...

Also, it looks like GS (via Wicket) is sending back JS (via ajax) that is then evaluated (executed) by wicket on the client end.  Not sure if this should be allowed or not...

I am going to look into this - saving this here for documentation purposes.

**Comment by davidblasby** [comment]
Hi, I did a quick experiment (module status):

image

The new ModalDialog has no header/title, or resize/move functionality out-of-the-box.

I can add that - but I am surprised I it doesn't already exist.  I looked around a bit for someone else's wicket extension, but didn't find anything obvious.  It might be better to use JQuery's dialog instead of the wicket dialog.  However, that's too many changes all over GS...
I will keep looking.  I don't think it will be too difficult to add JS/CSS functionality to it.

**Comment by davidblasby** [comment]
Ok, I did a bit more research on replacing the wicket `ModalWindow` with the new wicket `ModalDialog`.

1. The `ModalDialog` doesn't have very much functionality - as you can see from the sample picture (above) - thats all you get.  Its pretty much the simplest thing you could call a "dialog box". The old `ModalWindow` had a lot more functionality (i.e. drag, resize) and easier to use (title + pageCreators).

2. I looked through the GS code base, and the vast majority of the pop-up are pretty simple.  There's a few that aren't (i.e. in the wps input/subprocess popup, SRS chooser pop, and a few others), but they shouldnt be too difficult to modify.  I think the `ModalDialog` has a much simpler API (just set content), which means we'll have to fill in some functionality.

3. I think we could make a new `GSModalDialog` that would have a title bar, "x" close button, etc..

4. I **think** (not sure) its possible to ship some JS with this new `GSModalDialog` that would allow the drag/resize functionality.  I'm not 100% sure how to do this in the Wicket and `ModalDialog` workflow.  We would have to write this - shouldn't be too difficult.  There are some libraries for this, but I don't think we want to include them (long term maintainace/security of a JS library like jquery-ui).

5. After the above, we would have to go through each GS popup and change it to the new `GSModalDialog`.  This would almost certainly involve changing how the dialogs are built and the content injected.  I'm not quite sure how to do this, but I dont think it will be too difficult as most dialogs are pretty simple.

**Comment by jodygarnett** [comment]
I made a test plan, [Wicket Test Plan](https://docs.google.com/spreadsheets/d/1pQmncG4zxpgJnHxeI4myFfOBD17U2CIMcy59II4XAfo/edit?usp=sharing) and filled in what I could figure out from the comments on this PR.

**Comment by MichelGabriel** [comment]
I did a lot of testing, visually everything looks good, but when you open a dialog there is an error. And there's a lot of dialogs.

The test results are in the document @jodygarnett provided.

**Comment by davidblasby** [comment]
I've made some progress;

1. uses an ajax action to close the window ("X" button)
2. uses simple JS to drag the window 
3. will look at adding resizing...
4. This is currently only for simple popups (only tested on this page)
5. Doesn't have some of the functionality of the original ModalWindow (like initialSize, but this would be doable)
6. This will take a lot of testing...

https://github.com/user-attachments/assets/84c5786e-b9e8-40f7-a116-b83b40f042d5

**Comment by jodygarnett** [comment]
How do you feel about it David? Is the amount of upfront work still worth it in comprising on jquery-ui?

The video looks just fine. I guess things like internal scrolling of contents are similar as before as the work you are doing is around decorations / interactions but the structure of the DOM ends up similar to before.

> This will take a lot of testing...

What is needed for testing? Different browsers (or at least webkit based and Firefox...)

**Comment by davidblasby** [comment]
Ok, made some more progress.  I added resizing support.  It was more difficult than I was expecting (as it always is with CSS).  I had to change some of the styling to get it functional, but I think its working well.  The resize control is the pink box in the bottom right - we can change this to anything (likely invisible because its right where you would expect it to be and there's a cursor change).


https://github.com/user-attachments/assets/c78a04f3-00df-4b06-a572-b55c574fa863

I think the next steps will be to do a little refactor so this is in the right spot, then start looking at some of the other pop-up and see how they work.  If they set content explicitly, it should be pretty easy to transition.  I noticed some of them do lazysetup, and that might be a bit more difficult.  Will see...  Wish me luck!

**Comment by jodygarnett** [comment]
Wishing you more luck! And a great weekend.

@davidblasby / @MichelGabriel / anyone interested can we talk on Monday please? I would like to figure out if we can get this in for the upcoming release candidate.

It would be excellent to:
- Have the update to Wicket8 in order have the project somewhat close to our roadmap and spring-framework-6 end of year goals
- Get useful testing out of a release candidate (we can share the spreadsheet and ask the community to test community modules etc...)

Based on Monday chat I can volunteer to do the release candidate ... Friday?!

**Comment by aaime** [comment]
Glanced over the comments, nice progress. Two comments:

* I would suggest to make the resize control visible, e.g., I would not guess it was there unless the dialog had a border like a normal operating system window. Looking at the text area I'm typing into... something like this would help to figure it's there:

image

* The branch does not compile, it needs to get a merge from main due to the ZoomContext changes.

**Comment by davidblasby** [comment]
I will see if I can find an icon..

I am using a tiny bit of jquery - it’s jquery-ui I was looking at (a different library, which I don’t think is really maintained and likely have security issues).

**Comment by aaime** [comment]
Something like this? Pick the license you prefer, it's done by yours truly

Resize Handle Small

**Comment by davidblasby** [comment]
ok, here is a demo of what I got done today.

It was much more complex than I was first thinking.  I also found some bugs (in 2.25.3) that I've also fixed.


https://github.com/user-attachments/assets/c8a8c1c4-400f-4ebd-b30b-422d9f515e2e

**Comment by davidblasby** [comment]
I started on the WPS Builder.  However, I'm a bit stuck on the "subprocess" editor.

1. Its quite complex - and I'm having issues with it (probably my lack of deep wicket knowledge).
2. It doesn't appear to be working in 2.25.3 (see video).
3. I'm pretty sure my GSModalDialog doesn't allow for multiple popups at the same time.
4. I'd recommend that we remove the "define/edit" popup for "subprocess" and just allow them to edit the XML in the already existing box.

https://github.com/user-attachments/assets/015ca979-e70a-4fe6-9e6d-b1984a84a6ba

@jodygarnett @aaime - what do you think of removing the popup?  Also, are there any other screens with multiple pop-ups?

**Comment by jodygarnett** [comment]
> what do you think of removing the popup?

Anything we need to do.

* context: WCS Builder and WPS Builder as "demos" and not mission critical. Which may be how they got broken and not noticed.

* bad idea: For the nested WPS Request Builder adding a *copy to clipboard"  would allow a manual way to use WPS Builder to define a subprocess, and then copy and paste it into another WPS Request Builder.

We may need to break eggs (demos) to make an omelette (spring-framework-6 update).

**Comment by jodygarnett** [comment]
> Also, are there any other screens with multiple pop-ups?

I was not aware that the WPS Builder had nested pop-ups. 

The only other place of similar complexity is the [WPS Security Access Control](https://docs.geoserver.org/latest/en/user/services/wps/security.html) pages where they get nested sub-pages that have an "edit/ok/canel" workflow like a dialog.  Rather than the usual "edit/save/apply/cancel" workflow.  Since those are pages and not dialogs I trust they are okay.

**Comment by aaime** [comment]
The subprocess builder has been broken for years, since the last Wicket upgrade. Feel free to remove it.

**Comment by davidblasby** [comment]
Based on feedback in the PMC Meeting, I have stopped work on the `GSModalDialog`.  I have created a PR (DO NOT MERGE) so the changes will not be lost.


https://github.com/geoserver/geoserver/pull/7871

**Comment by jodygarnett** [comment]
Okay @bradh we had a productive GeoServer PSC meeting today to sort out what we can do here to involve more people:

* @petersmythe started up a breakout room: https://app.gitter.im/#/room/#geoserver/wicket-9-upgrade:gitter.im
* we made a rough plan (you can see in the meeting notes)
  1. take brad fork wicket9 → shared wicket9 branch
  2. relax dialog deprecated thing
  3. relax CSP restrictions
  4. make a milestone release → results in a docker image
     * extensions can be tested using docker
     * community modules can be tested if we manually release
  5. ask for help doing the A/B testing

**Comment by jodygarnett** [comment]
Work moved to shared https://github.com/geoserver/geoserver/tree/wicket9 branch.

PR https://github.com/geoserver/geoserver/pull/7872

**Pull request closed without merging**